lib/model/CTokenListDataCategorizerBase.cc

```diff
@@ -385,43 +399,48 @@ void CTokenListDataCategorizerBase::acceptPersistInserter(
-    return std::bind(
-        static_cast<void (*)(const TTokenMIndex&, const TTokenListCategoryVec&, std::size_t, core::CStatePersistInserter&)>(
-            &CTokenListDataCategorizerBase::acceptPersistInserter),
-        std::cref(m_TokenIdLookup), std::cref(m_Categories),
-        m_MemoryCategorizationFailures, std::placeholders::_1);
+    return [this](core::CStatePersistInserter& inserter) {
+        return CTokenListDataCategorizerBase::acceptPersistInserter(
+            m_TokenIdLookup, m_Categories, m_MemoryCategorizationFailures, inserter);
+    };
 }
 
 CDataCategorizer::TPersistFunc CTokenListDataCategorizerBase::makeBackgroundPersistFunc() const {
-    return std::bind(
-        static_cast<void (*)(const TTokenMIndex&, const TTokenListCategoryVec&, std::size_t, core::CStatePersistInserter&)>(
-            &CTokenListDataCategorizerBase::acceptPersistInserter),
-        // Do NOT add std::ref wrappers around these arguments - they MUST be
-        // copied for thread safety
-        m_TokenIdLookup, m_Categories, m_MemoryCategorizationFailures,
-        std::placeholders::_1);
+    // Do NOT change this to capture the member variables by
+    // reference - they MUST be copied for thread safety
+    return [
+        tokenIdLookup = m_TokenIdLookup, categories = m_Categories,
+        memoryCategorizationFailures = m_MemoryCategorizationFailures
+    ](core::CStatePersistInserter & inserter) {
+        return CTokenListDataCategorizerBase::acceptPersistInserter(
+            tokenIdLookup, categories, memoryCategorizationFailures, inserter);
+    };
 }
 
 void CTokenListDataCategorizerBase::addCategoryMatch(bool isDryRun,
                                                      const std::string& str,
                                                      std::size_t rawStringLen,
                                                      const TSizeSizePrVec& tokenIds,
                                                      const TSizeSizeMap& tokenUniqueIds,
-                                                     TSizeSizePrVecItr& iter) {
+                                                     TSizeSizePrVecItr iter) {
     m_Categories[iter->second].addString(isDryRun, str, rawStringLen, tokenIds, tokenUniqueIds);
 
     std::size_t& count{iter->first};
+    bool wasCountRare{this->isCategoryCountRare(count)};
     ++count;
+    ++m_TotalCount;
+    if (wasCountRare && this->isCategoryCountRare(count) == false) {
+        --m_NumRareCategories;
+    }
 
     // Search backwards for the point where the incremented count belongs
     auto swapIter = m_CategoriesByCount.end();
-    auto checkIter = iter;
-    while (checkIter != m_CategoriesByCount.begin()) {
-        --checkIter;
-        if (count <= checkIter->first) {
+    while (iter != m_CategoriesByCount.begin()) {
+        --iter;
+        if (count <= iter->first) {
             break;
         }
-        swapIter = checkIter;
+        swapIter = iter;
```

Comment:
Shouldn't `m_CategoriesByCount` be in descending count order, in which case don't you need to shuffle: `iter - 1 -> iter`, `iter - 2 -> iter - 1`, ... rather than just swap the start and end of the range?

Reply:
Yes, good point.  This was intended to be a no-op refactor to avoid having 3 iterator variables but I've messed it up. 🤦

Reply:
Ok now I'm wondering if this whole loop doesn't just simplify (since you are only adding one to the count):
```
if (iter != m_CategoriesByCount.begin() && count > (iter - 1)->first) {
    std::iter_swap(iter, iter - 1);
}
```
If you were to add more than one to the count you'd need the swap in the loop.

Reply:
Oh this is of course wrong since you need to jump over all categories whose count is equal: ignore.

Reply:
But interestingly you are reordering the ones which are equal. Maybe this is fine, but it is interesting. It still feels slightly surprising the swap isn't in the loop, but I guess if only the partial order matters it's fine and more efficient.